pkg/multiboot/internal/trampoline/trampoline_linux_amd64.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Trampoline sets machine to a specific state defined
+// by multiboot v1 spec and boots the final kernel.
+// https://www.gnu.org/software/grub/manual/multiboot/multiboot.html#Machine-state.
+package trampoline
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/u-root/u-root/pkg/ubinary"
+)
+
+const (
+	trampolineBeginLow  = "u-root-trampoline"
+	trampolineBeginHigh = "-begin"
+	trampolineEnd       = "u-root-trampoline-end"
+
+	trampolineEntry = "u-root-entry-long"
+	trampolineInfo  = "u-root-info-long"
+)
+
+var trampolineBegin []byte
+
+var alwaysFalse bool
+
+func start()
+
+func init() {
+	// Cannot use "u-root-trampoline-begin" directly, because then there
+	// would be two locations of "u-root-trampoline-begin" sequence.
+	trampolineBegin = append([]byte(trampolineBeginLow), []byte(trampolineBeginHigh)...)
+
+	if alwaysFalse {
+		// Can never happen, but still need it for linker to include assembly to a binary.
+		start()
+	}
+
+	// Hope Go compiler will never get smart enough to optimize this function.
+}
+
+// alignUp aligns x to a 0x10 bytes boundary.
+// go compiler aligns TEXT parts at 0x10 bytes boundary.
+func alignUp(x int) int {
+	const mask = 0x10 - 1
+	return (x + mask) & ^mask
+}
+
+// Setup scans file for trampoline code and sets
+// values for multiboot info address and kernel entry point.
+func Setup(path string, infoAddr, entryPoint uintptr) ([]byte, error) {
+	d, err := extract(path)
+	if err != nil {
+		return nil, err
+	}
+	return patch(d, infoAddr, entryPoint)
+}
+
+// extract extracts trampoline segment from file.
+// trampoline segment begins after "u-root-trampoline-begin" byte sequence
+// and ends at "u-root-trampoline-end" byte sequence.
+func extract(path string) ([]byte, error) {
+	d, err := ioutil.ReadFile(path)
```

Comment:
Shouldn't this file already be mapped into memory?

Reply:
Also, if you use `ReadFile`, there are some cases where it might not find the trampoline. For example, if the package is dynamically loaded or upx is used.

Reply:
It should be, but I don't know how to access it.
I don't think that either ups or dynamical load is our case.

Reply:
Besides following the `start` pointer (which Go didn't seem to let you do), the only way I could think of is looking through /proc/self/maps. What you already have right now is definitely better.

I'll just keep a mental note of compression/dynamic loading for the off chance it comes up...